mindocr/models/necks/rpn/proposal_generator.py

```diff
@@ -0,0 +1,139 @@
+        nms_thresh (float): Threshold in NMS. default 0.5
+        min_size (flaot): Remove predicted boxes with either height or
+             width < min_size. default 0.1
+    """
+
+    def __init__(self, pre_nms_top_n=12000, post_nms_top_n=2000, nms_thresh=0.5, min_size=1.0):
+        super(ProposalGenerator, self).__init__()
+        self.pre_nms_top_n = pre_nms_top_n
+        self.post_nms_top_n = post_nms_top_n
+        self.min_size = min_size
+        self.topk = ops.TopK()
+        self.nms = ops.NMSWithMask(nms_thresh)
+
+    def construct(self, scores, bbox_deltas, anchors, im_shape):
+        N = scores[0].shape[0]
+        B = anchors[0].shape[-1]
+        pred_objectness_logits, pred_anchor_deltas = (), ()
+
+        for score in scores:
+            pred_objectness_logits = pred_objectness_logits + (score.transpose((0, 2, 3, 1)).reshape(N, -1),)
+        for delta in bbox_deltas:
+            pred_anchor_deltas = pred_anchor_deltas + (delta.transpose((0, 2, 3, 1)).reshape((N, -1, B)),)
+        rpn_rois, rpn_rois_mask = self.predict_proposals(anchors, pred_objectness_logits, pred_anchor_deltas, im_shape)
+        rpn_rois = ops.stop_gradient(rpn_rois)
+        rpn_rois_mask = ops.stop_gradient(rpn_rois_mask)
+        return rpn_rois, rpn_rois_mask
+
+    def predict_proposals(self, anchors, pred_objectness_logits, pred_anchor_deltas, image_sizes):
+        pred_proposals = self.decode_proposals(anchors, pred_anchor_deltas)
+        return self.find_top_rpn_proposals(pred_proposals, pred_objectness_logits, image_sizes)
+
+    def decode_proposals(self, anchors, pred_anchor_deltas):
+        """decode pred_anchor_deltas to true box xyxy"""
+        proposals = ()
+        for anchors_i, pred_anchor_deltas_i in zip(anchors, pred_anchor_deltas):
+            N = pred_anchor_deltas_i.shape[0]
+            B = anchors_i.shape[-1]
+            pred_anchor_deltas_i = pred_anchor_deltas_i.reshape(-1, B)
+            anchors_i = ops.tile(ops.expand_dims(anchors_i, 0), (N, 1, 1)).reshape(-1, B)
+            proposals_i = delta2bbox(
+                pred_anchor_deltas_i, anchors_i, weights=(1.0, 1.0, 1.0, 1.0)
+            )
+            proposals = proposals + (proposals_i.reshape(N, -1, B),)
+        return proposals
+
+    def find_top_rpn_proposals(self, proposals, pred_objectness_logits, image_sizes):
+        """get top post_nms_top_n proposals"""
+        # nms实现方式和torch有精度差异，待替换接口
+        # 1. Select top-k anchor after nms for every level and every image
+        boxes = []
+        for level_id, (proposals_i, logits_i) in enumerate(zip(proposals, pred_objectness_logits)):
+            batch_size, Hi_Wi_A, B = proposals_i.shape
+            # temp_proposals = ops.concat(
+            #     (ops.zeros((Hi_Wi_A, 2), proposals_i.dtype), ops.ones((Hi_Wi_A, 2), proposals_i.dtype)), axis=-1
+            # )
+            batch_boxes = []
+            # logits_i = ops.sigmoid(logits_i)
+            for b in range(batch_size):
```

Comment:
consider implement in matrix format to improve efficiency. ops.topk supports batch-level computing.